packages/loot-core/src/server/reports/app.ts

```diff
@@ -0,0 +1,111 @@
+import { v4 as uuidv4 } from 'uuid';
+
+import {
+  type CustomReportData,
+  type CustomReportEntity,
+} from '../../types/models';
+import { parseConditionsOrActions } from '../accounts/transaction-rules';
+import { createApp } from '../app';
+import * as db from '../db';
+import { requiredFields } from '../models';
+import { mutator } from '../mutators';
+import { undoable } from '../undo';
+
+import { ReportsHandlers } from './types/handlers';
+
+const reportModel = {
+  validate(report: CustomReportEntity, { update }: { update?: boolean } = {}) {
+    requiredFields('reports', report, ['conditions'], update);
+
+    if (!update || 'conditionsOp' in report) {
+      if (!['and', 'or'].includes(report.conditionsOp)) {
+        throw new Error('Invalid filter conditionsOp: ' + report.conditionsOp);
+      }
+    }
+
+    return report;
+  },
+
+  toJS(row: CustomReportData) {
+    return {
+      ...row,
+      conditionsOp: row.conditions_op,
+      filters: parseConditionsOrActions(row.conditions),
+    };
+  },
+
+  fromJS(report: CustomReportEntity) {
+    const { filters, conditionsOp, ...row }: CustomReportData = report;
+    if (conditionsOp) {
+      row.conditions_op = conditionsOp;
+      row.conditions = filters;
+    }
+    return row;
+  },
+};
+
+async function reportNameExists(
+  name: string,
+  reportId: string,
+  newItem: boolean,
+) {
+  const idForName: { id: string } = await db.first(
+    'SELECT id from reports WHERE tombstone = 0 AND name = ?',
+    [name],
+  );
+
+  if (idForName === null) {
+    return false;
+  }
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏can `idForName` be `null`? According to the type defined above it will always have a value.

Reply:
This was needed because of the order of my previous poorly nested statements. But with the updates you made we can remove it. Thanks!